x/stake/keeper/validator.go

```diff
@@ -199,78 +199,101 @@ func (k Keeper) ClearTendermintUpdates(ctx sdk.Context) {
+	validator.BondHeight, validator.BondIntraTxCounter = k.getBondDetails(ctx, oldFound, oldValidator, validator)
+	valPower := k.updateValidatorPower(ctx, oldFound, oldValidator, validator, pool)
+	cliffPower := k.GetCliffValidatorPower(ctx)
+
+	switch {
+	// if already bonded and power increasing only need to update tendermint
+	case powerIncreasing && !validator.Revoked &&
+		(oldFound && oldValidator.Status() == sdk.Bonded):
 
-	// retrieve the old validator record
-	oldValidator, oldFound := k.GetValidator(ctx, ownerAddr)
+		bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
+		store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
+
+	// if is a new validator and the new power is less than the cliff validator
+	case cliffPower != nil && !oldFound &&
+		bytes.Compare(valPower, cliffPower) == -1: //(valPower < cliffPower
+		// skip to completion
+
+	// if was unbonded and the new power is less than the cliff validator
+	case cliffPower != nil &&
+		(oldFound && oldValidator.Status() == sdk.Unbonded) &&
+		bytes.Compare(valPower, cliffPower) == -1: //(valPower < cliffPower
+		// skip to completion
+
+	default:
+		// update the validator set for this validator
+		updatedVal := k.UpdateBondedValidators(ctx, validator)
+		if updatedVal.Owner != nil { // updates to validator occurred  to be updated
+			validator = updatedVal
+		} else {
 
-	if validator.Revoked && oldValidator.Status() == sdk.Bonded {
-		validator = k.unbondValidator(ctx, validator)
+			// if decreased in power but still bonded, update Tendermint validator
+			// (if updatedVal is set, the validator has changed bonding status)
+			stillBonded := oldFound && oldValidator.Status() == sdk.Bonded
+			if stillBonded && oldValidator.PoolShares.Bonded().GT(validator.PoolShares.Bonded()) {
+				bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
+				store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
+			}
+		}
+	}
+
+	k.SetValidator(ctx, validator)
+	return validator
+}
+
+func (k Keeper) updateForRevoking(ctx sdk.Context, oldFound bool, oldValidator, newValidator types.Validator) types.Validator {
+	if newValidator.Revoked && oldFound && oldValidator.Status() == sdk.Bonded {
+		newValidator = k.unbondValidator(ctx, newValidator)
 
 		// need to also clear the cliff validator spot because the revoke has
 		// opened up a new spot which will be filled when
 		// updateValidatorsBonded is called
 		k.clearCliffValidator(ctx)
 	}
+	return newValidator
+}
 
-	powerIncreasing := false
-	if oldFound && oldValidator.PoolShares.Bonded().LT(validator.PoolShares.Bonded()) {
-		powerIncreasing = true
+func (k Keeper) getPowerIncreasing(ctx sdk.Context, oldFound bool, oldValidator, newValidator types.Validator) bool {
+	if oldFound && oldValidator.PoolShares.Bonded().LT(newValidator.PoolShares.Bonded()) {
+		return true
 	}
+	return false
+}
+
+func (k Keeper) getBondDetails(ctx sdk.Context, oldFound bool, oldValidator,
```

Comment:
This seems to actually modify state (counter+1). Should the name or at least a godoc reflect that?

Reply:
unexposed functions are not documented with godoc, this is just a lil helper function - still, updated to have a comment and a more accurate name (`bondIncrement`)